cmd/csi-provisioner/csi-provisioner.go

```diff
@@ -110,17 +129,104 @@ func init() {
+	classInformer := func() cache.SharedIndexInformer {
+		if utilversion.MustParseSemantic(serverVersion.GitVersion).AtLeast(utilversion.MustParseSemantic(defaultServerVersion)) {
+			return informerFactory.Storage().V1().StorageClasses().Informer()
+		}
+		return informerFactory.Storage().V1beta1().StorageClasses().Informer()
+	}()
+	secretInformer := informerFactory.Core().V1().Secrets().Informer()
+
 	// Create the provisioner: it implements the Provisioner interface expected by
 	// the controller
-	csiProvisioner := ctrl.NewCSIProvisioner(clientset, *csiEndpoint, *connectionTimeout, identity, *volumeNamePrefix, *volumeNameUUIDLength, grpcClient)
+	csiProvisioner := ctrl.NewCSIProvisioner(
+		clientset,
+		*csiEndpoint,
+		*connectionTimeout,
+		identity,
+		*volumeNamePrefix,
+		*volumeNameUUIDLength,
+		grpcClient,
+		secretInformer.GetStore(),
+		classInformer.GetStore(),
+	)
 	provisionController = controller.NewProvisionController(
 		clientset,
 		*provisioner,
 		csiProvisioner,
 		serverVersion.GitVersion,
+		controller.ClaimsInformer(claimInformer),
+		controller.VolumesInformer(volumeInformer),
+		controller.ClassesInformer(classInformer),
 	)
+
+	recorder := makeEventRecorder(clientset)
+
+	id, err := os.Hostname()
+	if err != nil {
+		glog.Fatalf("failed to get hostname: %v", err)
+	}
+
+	resourceLock, err = resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		"default",
+		*provisioner,
+		clientset.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		},
+	)
+	if err != nil {
+		glog.Fatalf("error creating lock: %v", err)
+	}
+}
+
+func makeEventRecorder(clientset kubernetes.Interface) record.EventRecorder {
+	eventBroadcaster := record.NewBroadcaster()
+	Recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: *provisioner})
+	eventBroadcaster.StartLogging(glog.Infof)
+	if clientset != nil {
+		glog.V(4).Infof("Sending events to api server.")
+		eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(clientset.CoreV1().RESTClient()).Events(Namespace)})
+	} else {
+		glog.Warning("No api server defined - no events will be sent to API server.")
+		return nil
+	}
+	return Recorder
 }
 
 func main() {
-	provisionController.Run(wait.NeverStop)
+
+	run := func(stopCh <-chan struct{}) {
+		go informerFactory.Start(stopCh)
+		synced := informerFactory.WaitForCacheSync(stopCh)
+		for tpy, sync := range synced {
+			if !sync {
+				glog.Errorf("Wait for shared cache %s sync timeout", tpy)
+				return
+			}
+		}
+
+		provisionController.Run(stopCh)
+	}
+
+	leaderelection.RunOrDie(leaderelection.LeaderElectionConfig{
+		Lock:          resourceLock,
+		LeaseDuration: 15 * time.Second,
+		RenewDeadline: 10 * time.Second,
+		RetryPeriod:   2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				glog.Fatalf("leader election lost")
+			},
+		},
+	})
+
+	panic("unreached")
```

Comment:
If this panic should never be reached, update the string to say so i.e. `panic("panic reached, leader election failed")`.

Reply:
I have deleted the test code.